src/util/sighash.rs

```diff
@@ -0,0 +1,904 @@
+// Rust Bitcoin Library
+// Written in 2021 by
+//   The rust-bitcoin developers
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! # Generalized, efficient, signature hash implementation
+//!
+//! Implementation of the algorithm to compute the message to be signed according to [Bip341](https://github.com/bitcoin/bips/blob/150ab6f5c3aca9da05fccc5b435e9667853407f4/bip-0341.mediawiki),
+//! [Bip143](https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki)
+//! and legacy
+//!
+
+pub use blockdata::transaction::SigHashType as LegacySigHashType;
+use consensus::{encode, Encodable};
+use core::fmt;
+use core::ops::{Deref, DerefMut};
+use hashes::{sha256, sha256d, Hash};
+use io;
+use util::taproot::{TapLeafHash, TapSighashHash};
+use SigHash;
+use {Script, Transaction, TxOut};
+
+use prelude::*;
+
+/// Efficiently calculates signature hash message for legacy, segwit and taproot inputs.
+#[derive(Debug)]
+pub struct SigHashCache<T: Deref<Target = Transaction>> {
+    /// Access to transaction required for various introspection, moreover type
+    /// `T: Deref<Target=Transaction>` allows to accept borrow and mutable borrow, the
+    /// latter in particular is necessary for [SigHashCache::access_witness]
+    tx: T,
+
+    /// Common cache for taproot and segwit inputs. It's an option because it's not needed for legacy inputs
+    common_cache: Option<CommonCache>,
+
+    /// Cache for segwit v0 inputs, it's the result of another round of sha256 on `common_cache`
+    segwit_cache: Option<SegwitCache>,
+
+    /// Cache for taproot v1 inputs
+    taproot_cache: Option<TaprootCache>,
+}
+
+/// Values cached common between segwit and taproot inputs
+#[derive(Debug)]
+struct CommonCache {
+    prevouts: sha256::Hash,
+    sequences: sha256::Hash,
+
+    /// in theory, `outputs` could be `Option` since `NONE` and `SINGLE` doesn't need it, but since
+    /// `ALL` is the mostly used variant by large, we don't bother
+    outputs: sha256::Hash,
+}
+
+/// Values cached for segwit inputs, it's equal to [CommonCache] plus another round of `sha256`
```

Comment:
nit: doc comments should use ``[`CommonCache`]`` style (the same applies to the above deprecated messages

Reply:
Using backtick for this and for other type references in the file in c704ee7ffefbcc5fda2ef0ab3967eb9c7842865f

Reply:
Another nit: I think it's `CommonCache` that has the extra round of sha256, and `SegwitCache` has only one

Reply:
Oh, no, I am confusing segwit v1 with v0 aren't I. Never mind.